Add reference architecture page (#622)
apollographql/docs: src/content/graphos/enterprise/reference-architecture.mdx

---
title: Reference Architecture
description: Reference for Enterprise Deployment of GraphOS
---

<EnterpriseFeature>

Self-hosting the Apollo Router is limited to [GraphOS Enterprise plans](/graphos/enterprise/). Other plan types use [managed cloud routing with GraphOS](/graphos/routing/cloud/). Check out the [pricing page](https://www.apollographql.com/pricing/) to learn more.
</EnterpriseFeature>

Apollo supplies a reference architecture that is a blueprint to self-hosting the router and subgraphs in an enterprise cloud environment using Kubernetes and Helm.

It consists of the following Github repositories:

- [build-a-supergraph](https://github.com/apollosolutions/build-a-supergraph) - The main repository that contains the step-by-step guide to utilizing the architecture and deploying it to AWS or GCP

> - [build-a-supergraph-infra](https://github.com/apollosolutions/build-a-supergraph-infra) - The template repository for Kubernetes deployment of the router, collector, Grafana, performance tests, and Zipkin

Reviewer: Updated this to OTel collector. Correct if that's inaccurate.

| ------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------ |
| [build-a-supergraph](https://github.com/apollosolutions/build-a-supergraph)                       | The main repository that contains a step-by-step guide to utilizing the architecture and deploying it to AWS or GCP            |
| [build-a-supergraph-infra](https://github.com/apollosolutions/build-a-supergraph-infra)           | The template repository for Kubernetes deployment of the Apollo Router, OTel collector, Grafana, performance tests, and Zipkin |
| [build-a-supergraph-subgraph-a](https://github.com/apollosolutions/build-a-supergraph-subgraph-a) | Template repository for a subgraph used in the reference architecture                                                          |
| [build-a-supergraph-subgraph-b](https://github.com/apollosolutions/build-a-supergraph-subgraph-b) | Template repository for a subgraph used in the reference architecture                                                          |

## Getting started

> To get started with the reference architecture, begin with the README in the main [build-a-supergraph](https://github.com/apollosolutions/build-a-supergraph) repository. The README walks you through everything from setup to CI/CD to load testing the resulting architecture.

Author: Added a bit here but not sure if it's still addressing your feedback.

Open to suggestions here!

## Architecture overview

The reference architecture uses two Kubernetes clusters to support development and production environments. Each cluster has pods for:

- hosting the router
- hosting subgraphs
- trace collection
- K6 load testing with Grafana for viewing the results

For both environments, GraphOS serves as a schema registry. Each subgraph publishes schema updates to the registry via CI/CD. The router regularly polls an endpoint called [Apollo Uplink](/federation/managed-federation/uplink/) to get the latest schema and routing configurations from GraphOS.

```mermaid
graph LR;
  subgraph "Your infrastructure"
  serviceA[Subgraph A];
  serviceB[Subgraph B];
  router([Router]);
  end
  subgraph "Apollo GraphOS"
  registry{{Schema <br/> Registry}};
  uplink{{Apollo <br/> Uplink}}
  end
  serviceA & serviceB -->|"Publishes<br/>schema"| registry;
  registry -->|"Updates"| uplink;
  router -->|Polls for schema <br/> and configuration changes| uplink;
  class registry secondary;
  class uplink secondary;
```

The router also pushes performance and utilization metrics to GraphOS via Uplink so you can [analyze them in GraphOS Studio](../metrics/).

### Development environment

> ```mermaid

Author: Added some wording... please let me know if it makes sense!

flowchart TB
	developers[Developers]
	subgraph cloud["Cloud (AWS or GCP)"]
		subgraph devcluster["Dev Cluster"]
			subgraph ingresscontainer["Ingress"]
				ingress{"🔀"}
			end

			router["Router"]
			subgrapha["Subgraph A"]
			subgraphb["Subgraph B"]
		end
	end

  subgraph graphos["Apollo GraphOS"]
    uplink["Schema Registry/Uplink"]
    checks["Schema Checks"]
    metrics["Usage Metrics"]
  end

  subgraph github["GitHub"]
      githubactions["Subgraph CI/CD (GitHub Actions)"]
  end

	%% Relationships
	developers --> ingress
	ingress --> router
	router --> subgrapha
	router --> subgraphb
	uplink -.-> router
  router -.-> metrics
  githubactions -.-> uplink
  uplink -.-> checks

```

### Production environment

> 

Author: Added some wording... please let me know if it makes sense!

```mermaid
flowchart TB
	applications[Applications / Consumers]
	subgraph cloud["Cloud (AWS or GCP)"]
		subgraph devcluster["Prod Cluster"]
			subgraph ingresscontainer["Ingress"]
				ingress{"🔀"}
			end

			router["Router"]
			subgrapha["Subgraph A"]
			subgraphb["Subgraph B"]

			collector["OTel Collector"]
			zipkin[Zipkin]

			k6["K6 Load Tester"]
			grafana["Grafana"]
      influxdb[("InfluxDB")]

		end
	end

  subgraph graphos["Apollo GraphOS"]
    uplink["Schema Registry/Uplink"]
    checks["Schema Checks"]
    metrics["Usage Metrics"]
  end

  subgraph github["GitHub"]
      githubactions["Subgraph CI/CD (GitHub Actions)"]
  end

	%% Relationships
	applications --> ingress
	ingress --> router
	router --> subgrapha
	router --> subgraphb

	router -.-> collector
	subgrapha -.-> collector
	subgraphb -.-> collector
	collector -.-> zipkin

	k6 -.-> router
	k6 -.-> influxdb
  grafana --> influxdb

  uplink -.-> router
  router -.-> metrics
  githubactions -.-> uplink
  uplink -.-> checks
```

## Components

> 

Author: Took a crack at it but please let me know if you have any suggestions for different wording!

    collector["OTel Collector"]
    zipkin[Zipkin]
  end

  trigger -.-> deploy
  deploy --> collector
  deploy --> zipkin
```

### Deploy load test infrastructure

This job is manually triggered within GitHub Actions.

```mermaid
flowchart LR
  trigger(["Trigger Production Deployment"])

  subgraph githubactions["GitHub Actions"]
    deploy["Job: Deploy Load Test Infrastructure"]
  end

  subgraph cluster["Production Cluster"]
    k6["K6 Load Tester"]
		grafana["Grafana"]
		tests["Load Tests"]
    influxdb[("InfluxDB")]
  end

  trigger -.-> deploy
  deploy --> k6
  deploy --> grafana
  deploy --> tests
  deploy --> influxdb
```

### Run load tests

This job is manually triggered within GitHub Actions.

```mermaid
flowchart LR
  trigger(["Trigger Load Tests"])

  subgraph githubactions["GitHub Actions"]
    run["Job: Run Load Tests"]
  end

  subgraph cluster["Production Cluster"]
    router["Router"]
    subgraphs["Subgraphs"]

    k6["K6 Load Tester"]
		tests["Load Tests"]
    influxdb[("InfluxDB")]
  end

  router --> subgraphs

  trigger -.-> run
  run --> k6
  k6 --> router
  k6 -.-> tests
  k6 -.-> influxdb
```

## Further reading

- [Router Introduction](/router/)

> - [Enterprise Features](/router/enterprise-features)

Reviewer: While the other further reading seem explicitly related to this guide, I'm not sure the enterprise features one is necessary to add?

```suggestion
```

Author: My thought process of including it was to try to push readers towards understanding the enterprise components. This will become especially relevant as we start adding enterprise-only features to the reference architecture (which we have already started to do).

Happy to remove it though if it still doesn't make sense. :)

| [build-a-supergraph](https://github.com/apollosolutions/build-a-supergraph)                       | The main repository that contains a step-by-step guide to utilizing the architecture and deploying it to AWS or GCP            |
| [build-a-supergraph-infra](https://github.com/apollosolutions/build-a-supergraph-infra)           | The template repository for Kubernetes deployment of the Apollo Router, OTel collector, Grafana, performance tests, and Zipkin |
| [build-a-supergraph-subgraph-a](https://github.com/apollosolutions/build-a-supergraph-subgraph-a) | Template repository for a subgraph used in the reference architecture                                                          |
| [build-a-supergraph-subgraph-b](https://github.com/apollosolutions/build-a-supergraph-subgraph-b) | Template repository for a subgraph used in the reference architecture                                                          |

## Getting started

To get started with the reference architecture, begin with the README in the main [build-a-supergraph](https://github.com/apollosolutions/build-a-supergraph) repository. 

> The README walks you through everything from setup to CI/CD to load testing the resulting architecture. You can utilize this page as a reference for how the architecture
> is setup.

Reviewer: ```suggestion
The README walks you through building a supergraph with the reference architecture. It gives step-by-step instructions for setup, CI/CD,  load testing, and more. This page serves as a complementary reference of the architecture's organization and configuration.
```

## Architecture overview

The reference architecture uses two Kubernetes clusters to support development and production environments. Each cluster has pods for:

- hosting the router
- hosting subgraphs
- trace collection
- K6 load testing with Grafana for viewing the results

For both environments, GraphOS serves as a schema registry. Each subgraph publishes schema updates to the registry via CI/CD. The router regularly polls an endpoint called [Apollo Uplink](/federation/managed-federation/uplink/) to get the latest schema and routing configurations from GraphOS.

```mermaid
graph LR;
  subgraph "Your infrastructure"
  serviceA[Subgraph A];
  serviceB[Subgraph B];
  router([Router]);
  end
  subgraph "Apollo GraphOS"
  registry{{Schema <br/> Registry}};
  uplink{{Apollo <br/> Uplink}}
  end
  serviceA & serviceB -->|"Publishes<br/>schema"| registry;
  registry -->|"Updates"| uplink;
  router -->|Polls for schema <br/> and configuration changes| uplink;
  class registry secondary;
  class uplink secondary;
```

The router also pushes performance and utilization metrics to GraphOS via Uplink so you can [analyze them in GraphOS Studio](../metrics/).

> ### Development environment
> The development environment consists of a the router and its subgraphs hosted in a Kubernetes cluster in either AWS or GCP. The subgraph schemas
> are pushed to Apollo GraphOS via the Schema Uplink and verified using schema checks. The resulting supergraph schema is made available to the Router. The Router reports usage metrics back to GraphOS.
> 

Reviewer: Thanks for adding this!

Just a couple of tweaks:
- Whenever possible, use active rather than passive voice. It makes clear who is doing what and makes writing less verbose and more readable.
- Add links when relevant (e.g. for schema checks) so the reader can easily learn more
- See [content guidelines for the router](https://apollographql.quip.com/AOUaATYpDc0q/Apollo-Products-and-Features#IOPACAUxgwi) -> _In educational content, you may refer to as "the router" (uncapitalized) after a nearby preceding use of "the Apollo Router". Do not use "the Router" (capitalized)._

```suggestion
### Development environment

The development environment consists of the router and its subgraphs hosted in a Kubernetes cluster in either AWS or GCP.  GraphOS validates subgraph schemas using [schema checks](/federation/managed-federation/federated-schema-checks/) and makes them available via [Uplink](/federation/managed-federation/uplink/). The router can access the resulting supergraph schema through Uplink. The router also reports usage metrics back to GraphOS.

```

```mermaid
flowchart TB
	developers[Developers]
	subgraph cloud["Cloud (AWS or GCP)"]
		subgraph devcluster["Dev Cluster"]
			subgraph ingresscontainer["Ingress"]
				ingress{"🔀"}
			end

			router["Router"]
			subgrapha["Subgraph A"]
			subgraphb["Subgraph B"]
		end
	end

  subgraph graphos["Apollo GraphOS"]
    uplink["Schema Registry/Uplink"]
    checks["Schema Checks"]
    metrics["Usage Metrics"]
  end

  subgraph github["GitHub"]
      githubactions["Subgraph CI/CD (GitHub Actions)"]
  end

	%% Relationships
	developers --> ingress
	ingress --> router
	router --> subgrapha
	router --> subgraphb
	uplink -.-> router
  router -.-> metrics
  githubactions -.-> uplink
  uplink -.-> checks

```

### Production environment

> The production environment is similar to the development environment with some additions. 

Reviewer: ```suggestion

The production environment is similar to the development environment with some additions. 
```

> - The router and subgraphs send their OpenTelemetry data to a collector to be viewable in Zepkin

Reviewer: ```suggestion
- The router and subgraphs send their OpenTelemetry data to a collector. You can then view the data in Zipkin.
```

> - The Load Testing infrastructure (K6 Load Tester, InfluxDB, Grafana) are deployed. K6 is configured to send traffic to the router.

Reviewer: ```suggestion
- A K6 load tester sends traffic to the router and stores load test results in InfluxDB for viewing in Grafana.
```

```mermaid
flowchart TB
	applications[Applications / Consumers]
	subgraph cloud["Cloud (AWS or GCP)"]
		subgraph devcluster["Prod Cluster"]
			subgraph ingresscontainer["Ingress"]
				ingress{"🔀"}
			end

			router["Router"]
			subgrapha["Subgraph A"]
			subgraphb["Subgraph B"]

			collector["OTel Collector"]
			zipkin[Zipkin]

			k6["K6 Load Tester"]
			grafana["Grafana"]
      influxdb[("InfluxDB")]

		end
	end

  subgraph graphos["Apollo GraphOS"]
    uplink["Schema Registry/Uplink"]
    checks["Schema Checks"]
    metrics["Usage Metrics"]
  end

  subgraph github["GitHub"]
      githubactions["Subgraph CI/CD (GitHub Actions)"]
  end

	%% Relationships
	applications --> ingress
	ingress --> router
	router --> subgrapha
	router --> subgraphb

	router -.-> collector
	subgrapha -.-> collector
	subgraphb -.-> collector
	collector -.-> zipkin

	k6 -.-> router
	k6 -.-> influxdb
  grafana --> influxdb

  uplink -.-> router
  router -.-> metrics
  githubactions -.-> uplink
  uplink -.-> checks
```

## Components

> This section will summarize the various services and runtimess that make up the architecture. This includes the router, subgraphs, OpenTelemetry, Zipkin, K6, and Grafana.

Reviewer: ```suggestion
This section summarizes the various services and runtimes that make up the reference architecture. This includes the router, subgraphs, OpenTelemetry, Zipkin, K6, and Grafana.
```

	%% Relationships
	router --> subgrapha
	router --> subgraphb

	k6 ---> router
	k6 -.-> influxdb
  grafana --> influxdb
```

## CI/CD

The reference architecture uses [GitHub Actions](https://docs.github.com/en/actions) for its CI/CD. These actions include:

- PR-level schema checks
- Building containers using Docker
- Publishing subgraph schemas to Apollo Uplink
- Router deployment to the Kubernetes cluster
- OTel collector deployment
- Grafana deployment
- Running load tests

### Development actions

Submitting and merging a PR triggers the following GitHub Actions workflows:

```mermaid
flowchart TB
  submit(["Submit PR"])
  mergedpr(["Merge PR"])

  subgraph githubactions["GitHub Actions"]
    checkspr["Job: Schema Checks"]
    builddev["Job: Build, Deploy, Schema Publish"]
  end

  subgraph devcluster["Dev Cluster"]
    router["Router"]
    subgraphs["Subgraph"]
  end

  subgraph graphos["Apollo Graphos"]
    checks["Schema Checks"]
    uplink["Schema Uplink"]
  end

  submit -.-> checkspr
  checkspr --> checks
  submit --> mergedpr
  mergedpr -.-> builddev
  builddev --> subgraphs
  builddev --> uplink
  uplink -.-> checks
  uplink -.-> router
```

### Production deploy

> You must manually trigger production deployments with GitHub Actions.

Reviewer: Does the following suggestion accurately describe what's happening? The diagram is a little confusing, since it looks like GH actions publish changes to a subgraph, whereas previous diagrams/descriptions show the change originating in a subgraph.

```suggestion
When you manually trigger a production deployment, GitHub Actions publish schema updates to Uplink and GraphOS validates them using schema checks before making them available to the router.
```

Author: What is missing from this description, and what the diagram is supposed to show, is both that the schema is published to Uplink _and_ the subgraph services themselves are deployed.

```mermaid
flowchart TB
  trigger(["Trigger Production Deployment"])

  subgraph githubactions["GitHub Actions"]
    builddev["Job: Build, Deploy, Schema Publish"]
  end

  subgraph devcluster["Production Cluster"]
    router["Router"]
    subgraphs["Subgraph"]
  end

  subgraph graphos["Apollo Graphos"]
    checks["Schema Checks"]
    uplink["Schema Uplink"]
  end

  trigger -.-> builddev
  builddev --> subgraphs
  builddev --> uplink
  uplink -.-> checks
  uplink -.-> router
```

### Deploy router

> This job is manually triggered within GitHub Actions.

Reviewer: Could you please add similar descriptions for the following three mermaid diagrams?

```mermaid
flowchart LR
  trigger(["Trigger Deployment"])

  subgraph githubactions["GitHub Actions"]
    deploy["Job: Deploy Router"]
  end

  subgraph cluster["Dev/Production Cluster"]
    router["Router"]
  end

  trigger -.-> deploy
  deploy --> router
```

### Deploy OpenTelemetry collector

This job is manually triggered within GitHub Actions.

```mermaid
flowchart LR
  trigger(["Trigger Deployment"])

  subgraph githubactions["GitHub Actions"]

>     deploy["Job: Deploy OTEL"]

Reviewer: ```suggestion
    deploy["Job: Deploy OTel"]
```

  subgraph cluster["Production Cluster"]
    collector["OTel Collector"]
    zipkin[Zipkin]
  end

  trigger -.-> deploy
  deploy --> collector
  deploy --> zipkin
```

### Deploy load test infrastructure

This job is manually triggered within GitHub Actions.

```mermaid
flowchart LR
  trigger(["Trigger Production Deployment"])

  subgraph githubactions["GitHub Actions"]
    deploy["Job: Deploy Load Test Infrastructure"]
  end

  subgraph cluster["Production Cluster"]
    k6["K6 Load Tester"]
		grafana["Grafana"]
		tests["Load Tests"]
    influxdb[("InfluxDB")]
  end

  trigger -.-> deploy
  deploy --> k6
  deploy --> grafana
  deploy --> tests
  deploy --> influxdb
```

### Run load tests

This job is manually triggered within GitHub Actions.

```mermaid
flowchart LR
  trigger(["Trigger Load Tests"])

  subgraph githubactions["GitHub Actions"]
    run["Job: Run Load Tests"]
  end

  subgraph cluster["Production Cluster"]
    router["Router"]
    subgraphs["Subgraphs"]

    k6["K6 Load Tester"]
		tests["Load Tests"]
    influxdb[("InfluxDB")]
  end

  router --> subgraphs

  trigger -.-> run
  run --> k6
  k6 --> router
  k6 -.-> tests
  k6 -.-> influxdb
```

## Further reading

> - [Apollo Router introduction](/router/)
> - [Containerizing the Apollo Router](/router/containerization/overview)
> - [Managing Apollo Router resources in Kubernetes](/technotes/TN0016-router-resource-management/)
> - [Enterprise features of the Apollo Router](/router/enterprise-features)
> - [Sending Apollo Router operation metrics to GraphOS](/router/configuration/apollo-telemetry)
> - [OpenTelemetry tracing in the Apollo Router](/router/configuration/tracing)

Reviewer: I'm fine with keeping the Enterprise link, but it seems kind of randomly placed amongst the other links which generally follow the structure of the reference architecture. I think it would make more sense to put closer to the beginning (right after the router intro), or at the very end like this:

```suggestion
- [Apollo Router introduction](/router/)
- [Containerizing the Apollo Router](/router/containerization/overview)
- [Managing Apollo Router resources in Kubernetes](/technotes/TN0016-router-resource-management/)
- [Sending Apollo Router operation metrics to GraphOS](/router/configuration/apollo-telemetry)
- [OpenTelemetry tracing in the Apollo Router](/router/configuration/tracing)
- [Enterprise features of the Apollo Router](/router/enterprise-features)
```

---
title: Reference Architecture
description: Reference for Enterprise Deployment of GraphOS
---

<EnterpriseFeature>

While you can run the Apollo Router regardless of your Apollo plan, connecting the router to GraphOS requires an [Enterprise plan](https://new.apollographql.com/pricing). If your organization doesn't currently have an Enterprise plan, you can test out this functionality by signing up for a free [Enterprise trial](https://studio.apollographql.com/signup?type=enterprise-trial&referrer=docs-content).

</EnterpriseFeature>

> In a modern cloud-native stack, your components must be scalable with high availability. The [Apollo Router](/router) is built with this in mind. The router is [much faster and less resource-intensive than the Apollo Gateway](https://www.apollographql.com/blog/announcement/backend/apollo-router-our-graphql-federation-runtime-in-rust/#apollo-router-vs-apollo-gateway-benchmarks), Apollo's original runtime.

Reviewer 2: ```suggestion
In this guide, learn about the fundamental concepts and configuration underlying Apollo's reference architecture for enterprise deployment of GraphOS with a self-hosted router. Use this guide as a companion reference to the [Apollo reference architecture repositories](#reference-architecture-repositories).

## About Apollo's reference architecture

In a modern cloud-native stack, your components must be scalable with high availability. The [Apollo Router](/router) is built with this in mind. The router is [much faster and less resource-intensive than the Apollo Gateway](https://www.apollographql.com/blog/announcement/backend/apollo-router-our-graphql-federation-runtime-in-rust/#apollo-router-vs-apollo-gateway-benchmarks), Apollo's original runtime.
```

Apollo provides a reference architecture for self-hosting the router and subgraphs in an enterprise cloud environment using Kubernetes and Helm. This reference architecture includes autoscaling to suit the needs of a modern enterprise application. Continue reading to learn more about the reference architecture and how to use it.

<Tip>

Check out our blog post, [Deploying the Apollo Router at Apollo](https://www.apollographql.com/blog/announcement/platform/deploying-the-apollo-router-at-apollo/), to learn about Apollo's
internal use of the router, including the performance improvements and resource utilization reductions.

</Tip>

> ## Reference architecture structure

Reviewer 2: ```suggestion
## Reference architecture repositories
```

| ------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------ |
| [build-a-supergraph](https://github.com/apollosolutions/build-a-supergraph)                       | The main repository that contains a step-by-step guide to utilizing the architecture and deploying it to AWS or GCP            |
| [build-a-supergraph-infra](https://github.com/apollosolutions/build-a-supergraph-infra)           | The template repository for Kubernetes deployment of the Apollo Router, OTel collector, Grafana, performance tests, and Zipkin |
| [build-a-supergraph-subgraph-a](https://github.com/apollosolutions/build-a-supergraph-subgraph-a) | Template repository for a subgraph used in the reference architecture                                                          |
| [build-a-supergraph-subgraph-b](https://github.com/apollosolutions/build-a-supergraph-subgraph-b) | Template repository for a subgraph used in the reference architecture                                                          |

## Getting started

> To get started with the reference architecture, begin with the README in the main [build-a-supergraph](https://github.com/apollosolutions/build-a-supergraph) repository. 
> The README walks you through building a supergraph with the reference architecture. It gives step-by-step instructions for setup, CI/CD,  load testing, and more. This page serves as a complementary reference of the architecture's organization and configuration.

Reviewer 2: ```suggestion
To get started with the reference architecture, follow the README in the main [build-a-supergraph](https://github.com/apollosolutions/build-a-supergraph) repository. 
The README provides a how-to guide that walks you through building a supergraph with the reference architecture. It gives step-by-step instructions for setup, CI/CD, load testing, and more. 

While you work through the steps of deploying the reference architecture, you can use this guide as a complementary reference of the architecture's organization and configuration.
```

## Architecture overview

> The reference architecture uses two Kubernetes clusters to support development and production environments. Each cluster has pods for:

Reviewer 2: ```suggestion
The reference architecture uses two Kubernetes clusters, one for a development environment and other for a production environment. Each cluster has pods for:
```

> - hosting the router
> - hosting subgraphs
> - trace collection
> - K6 load testing with Grafana for viewing the results

Reviewer 2: ```suggestion
- Hosting the router
- Hosting subgraphs
- Collecting traces
- Load testing K6 and viewing results with Grafana
```

> For both environments, GraphOS serves as a schema registry. Each subgraph publishes schema updates to the registry via CI/CD. The router regularly polls an endpoint called [Apollo Uplink](/federation/managed-federation/uplink/) to get the latest schema and routing configurations from GraphOS.

Reviewer 2: ```suggestion
For both environments, GraphOS serves as a schema registry. Each subgraph publishes schema updates to the registry via CI/CD, and GraphOS validates and composes them into a supergraph schema. The router regularly polls an endpoint called [Apollo Uplink](/federation/managed-federation/uplink/) to get the latest supergraph schema and routing configurations from GraphOS.
```

```yaml [title="_run-loadtest-aws.yaml"]
spec:
  parallelism: ${{ inputs.parallelism }}
  arguments: '--out influxdb=http://influxdb.monitoring:8086/db'
  script:
    configMap:
      name: tests
      file: ${{ inputs.test }}.js
```

```yaml [title="value.yaml"]
datasources:
  datasources.yaml:
    apiVersion: 1
    datasources:
      - name: InfluxDB
        type: influxdb
        access: proxy
        url: http://influxdb.monitoring:8086
```

```mermaid
flowchart LR
  subgraph devcluster["Prod Cluster"]

    router["Router"]
    subgrapha["Subgraph A"]
    subgraphb["Subgraph B"]

    k6["K6 Load Tester"]
    influxdb[("InfluxDB")]
    grafana["Grafana"]

  end

	%% Relationships
	router --> subgrapha
	router --> subgraphb

	k6 ---> router
	k6 -.-> influxdb
  grafana --> influxdb
```

## CI/CD

The reference architecture uses [GitHub Actions](https://docs.github.com/en/actions) for its CI/CD. These actions include:

- PR-level schema checks
- Building containers using Docker
- Publishing subgraph schemas to Apollo Uplink
- Router deployment to the Kubernetes cluster
- OTel collector deployment
- Grafana deployment
- Running load tests

### Development actions

> When a PR is submitted, Github Actions uses GraphOS to validate them using schema checks.

Reviewer: ```suggestion
When a PR is submitted, GitHub Actions uses GraphOS to validate schema changes using schema checks.
```

> When a PR is submitted, GitHub Actions uses GraphOS to validate schema changes using schema checks.

Reviewer 2: For which repos does this apply to, all of them or the main?
